contents/scala/Scala.md

```diff
@@ -0,0 +1,241 @@
+
+Classes are extended by *subclassing* and a flexible *mixin-based* mechanism to avoid the problems of multiple inheritance.
+
+Classes cannot have *static members*. A singleton object with same name of the class can be used to achieve the same effect. 
+A singleton object is basically a class that can have only one instance.
+
+### Functional
+Scala is a **Functional Programming language** in the sense that every function in Scala is a value. 
+
+Scala provides a lightweight syntax for defining [*anonymous functions*](https://docs.scala-lang.org/tour/basics.html#functions).
+
+It also support many other features of FP including [*higher-order functions*](https://docs.scala-lang.org/tour/higher-order-functions.html), 
+[*nested functions*](https://docs.scala-lang.org/tour/nested-functions.html), [*currying*](https://docs.scala-lang.org/tour/multiple-parameter-lists.html), 
+[*pattern matching*](https://docs.scala-lang.org/tour/pattern-matching.html), *immutability* and *lazy evaluation*.
+
+# Getting Started
+There are mainly two ways to work in Scala.
+
+1. Using the **INTELLIJ IDE**. Refer to the [instructions](https://docs.scala-lang.org/getting-started-intellij-track/getting-started-with-scala-in-intellij.html) for more details.
+2. Using the **command line**. Refer to the [instructions](https://docs.scala-lang.org/getting-started-sbt-track/getting-started-with-scala-and-sbt-on-the-command-line.html) for more details.
+
+# Why Scala
+There are mainly two benefits of Scala, **simple syntax** and **multi-paradigm**. 
+
+*Simple syntax* allows us to write more simple and elegant code. 
+
+*Multi-paradigm* allows us to use an OOP language while using the feature in functional programming. 
+
+Following are some examples to illustrate the benefit of Scala:
+
+## Operator Overloading
+All operators in Scala such as +, -, *, /. are actually methods.
+
+Normally, we write addition in following way:
+
+`val a = 1 + 2`
+
+However, in Scala, it will be interpreted as following:
+
+`val a = 1.+(2)`
+
+Knowing that, we are able to overload operators just like method overloading. It allows us to redefine the behavior of
+arithmetic operators and give them meaning for our own custom classes.
+
+Following is an elegant *ComplexInt* class which uses **operator overloading**.
+
+```
+case class ComplexInt(real: Int, im: Int) {
+  def + (other: ComplexInt) = ComplexInt(real + other.real, im + other.im)
+  def * (other: ComplexInt) = ComplexInt(
+    real * other.real - im * other.im,
+    real * other.im + im * other.real
+  )
+  def unary_- = ComplexInt(-real, -im)
+  def - (other: ComplexInt) = this + -other
+
+  override def toString = real + " + " + im + "i"
+}
+```
+
+We are able to use the class like below:
+
+```
+val a = ComplexInt(1, 1)
+val b = ComplexInt(1, -1)
+val c = a * b
+val d = -c + a
+```
+
+In other languages like Java, the similar operations of the same class may need to performed like following:
+
```

Comment:
I think it's better to put the Java at beginning, and say something like "You can only write new methods to handle computation, which may cause XXX problems". Scala supports operator overloading that can solve them. This could be more impressive